### DISCUSSION TITLE
TypeScript definitions not working

### DISCUSSION THREAD

**Discussion by HunterBarclay** [jrouwe/JoltPhysics.js#121] [General]
Whenever I try to use the typescript types, I can't explicitly state the type of a variable or use the type in any way. Even with the closest I've gotten, it just says its missing prototype functions. Has anyone else been able to successfully explicitly type variables?

**Comment by PhoenixIllusion** [comment]
Do you have a snippet you can show where you are attempting to type and it is failing? I personally code almost exclusively in typescript except when porting my demos back to the VanillaJS for examples or asking assistance on this discussion forum.

For stock typescript (TSC) should only be required to import the Jolt namespace from the NPM package to access the types.

```typescript
import Jolt from 'jolt-physics'

export type JVec3 = Jolt.Vec3 | Jolt.RVec3;

export const SetJoltVec3 = (vec3: Vector3, jVec3: Jolt.Vec3) => {
  jVec3.Set(vec3.x, vec3.y, vec3.z)
  return jVec3;
}
export const SetJoltRVec3 = (vec3: Vector3, jVec3: Jolt.RVec3) => {
  jVec3.Set(vec3.x, vec3.y, vec3.z)
  return jVec3;
}

export const GetJoltVec3 = (jVec3: JVec3, vec3: Vector3) => {
  vec3.set(jVec3.GetX(), jVec3.GetY(), jVec3.GetZ())
  return vec3;
}

```
I use it more in the following:
https://github.com/PhoenixIllusion/babylonjs-jolt-physics-plugin/blob/main/src/jolt-raycast.ts

There are Vite specific issues that I had to work around regarding Vite's packing caching and the package.json declarations, and I do not know if those extend to other ESBuild based compiles, since esbuild rolls its own typescript compiling.

I worked around this EsBuild issue by importing the actual library in a JS file and then adding my own *.d.ts definition, and also modified the corresponding JS file so that it supporting supplying arbitrary Jolt modules so projects using the typescript could supply their own via loading from an external url or their local host or Wasm-compat or Wasm file.

https://github.com/PhoenixIllusion/babylonjs-jolt-physics-plugin/blob/main/src/jolt-import.d.ts

**Comment by HunterBarclay** [comment]
I was able to figure my situation out. I'm working on off of a fork that has some modifications that made using explicit types more difficult. Thanks for the help

**Comment by dxu** [comment]
@PhoenixIllusion can I ask what the Vite specific issues that you had to deal with were? I'm running into an issue right now where the default `import Jolt from "jolt-physics"` aren't working, but if I go into my node modules and change the 

```
import Jolt from "./types";

export default Jolt;
export * from "./types";
```
to 
```
import Jolt from "./types.d.ts";

export default Jolt;
export * from "./types.d.ts";
```

and add the `.d.ts` (or just `.ts`), it magically resolves properly. I'm using vite and configuring typescript for ESM modules

**Reply by PhoenixIllusion** [reply]
My issue was with vite@4.5 , which I believe was updated to 5.0 beginning of last December.
It looks like any build higher than 5.0.0 has a fix, so any projects created since then would have it fixed out-of-the box, and any older projects can just upgrade to the newer ViteJS.

The old version did this.
```sh
#!/bin/bash
set -e

npm --yes create vite@latest typescript-jolt -- --template vanilla-ts
cd typescript-jolt
npm install
npm install vite@4.5
npm install jolt-physics
echo "import initJolt from 'jolt-physics'" > src/main.ts
echo "initJolt();" >> src/main.ts
npm run dev
```
Looking in the browser:
```
[plugin:vite:asset-import-meta-url] Failed to resolve entry for package 
"/workspace/typescript-jolt/node_modules/.vite/deps". The package may have incorrect
 main/module/exports specified in its package.json.
```
